File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / orbitalPeriod
  where orbitalPeriod = 31557600 * multiplierOn planet
        multiplierOn Earth   = 1
        multiplierOn Mercury = 0.2408467
        multiplierOn Venus   = 0.61519726
        multiplierOn Mars    = 1.8808158
        multiplierOn Jupiter = 11.862615
        multiplierOn Saturn  = 29.447498
        multiplierOn Uranus  = 84.016846
        multiplierOn Neptune = 164.79132

File: fsharp/space-age/SpaceAge.fs
﻿module SpaceAge

type Planet = Earth | Mercury | Venus | Mars | Jupiter | Saturn | Uranus | Neptune

let age (planet: Planet) (seconds: int64): float =
  let years = float(seconds) / (365.25 * 24.0 * 60.0 * 60.0)
  let multiplier =
    match planet with
    | Earth   -> 1.0
    | Mercury -> 0.2408467
    | Venus -> 0.61519726
    | Mars -> 1.8808158
    | Jupiter -> 11.862615
    | Saturn -> 29.447498
    | Uranus -> 84.016846
    | Neptune -> 164.79132
  years / multiplier
